### PR TITLE
Perception: reactivate traffic light detection component

### PR DESCRIPTION
**Pull request by xiaoMrzhang**

### CODE REVIEW

File: modules/perception/onboard/component/trafficlights_perception_component.cc [hunk 1]
```diff
@@ -20,6 +20,7 @@
 #include <utility>
 
 #include <boost/algorithm/string.hpp>
+#include <opencv2/opencv.hpp>
```
Comment (jeroldchen) [inline modules/perception/onboard/component/trafficlights_perception_component.cc:23]: Adjust headers' order.

File: modules/perception/onboard/component/trafficlights_perception_component.cc [hunk 2]
```diff
@@ -36,6 +37,8 @@
 #include "modules/perception/onboard/common_flags/common_flags.h"
 #include "modules/transform/proto/transform.pb.h"
 
+using ::apollo::cyber::Clock;
```
Comment (jeroldchen) [inline modules/perception/onboard/component/trafficlights_perception_component.cc:40]: Why not put this inside namespace?

Comment (lx18233184051) [reply modules/perception/onboard/component/trafficlights_perception_component.cc:40]: +1

File: modules/perception/onboard/component/trafficlights_perception_component.cc [hunk 3]
```diff
@@ -83,6 +86,8 @@ static int GetGpuId(
 }
 
 bool TrafficLightsPerceptionComponent::Init() {
+  std::shared_ptr<camera::CameraFrame> p(new camera::CameraFrame);
+  frame_.swap(p);
```
Comment (lx18233184051) [inline modules/perception/onboard/component/trafficlights_perception_component.cc:90]: why not make_shared?

File: modules/perception/onboard/component/trafficlights_perception_component.cc (hunk 1 above)
Comment (lx18233184051) [reply modules/perception/onboard/component/trafficlights_perception_component.cc:23]: this is ok？

File: modules/perception/onboard/component/trafficlights_perception_component.cc (hunk 3 above)
Comment (xiaoMrzhang) [reply modules/perception/onboard/component/trafficlights_perception_component.cc:90]: make_shared may cause Eigen error  @lfcarol

Comment (lx18233184051) [reply modules/perception/onboard/component/trafficlights_perception_component.cc:90]: or just frame_ = std::shared_ptr<camera::CameraFrame> (new camera::CameraFrame);?
or frame_.reset(new camera::CameraFrame );
:)

Comment (jinghaomiao) [reply modules/perception/onboard/component/trafficlights_perception_component.cc:91]: It's a bit awkard to instantiate a new shared pointer and swap with an existing one. Why not reset it to the existing value or instantialize it to the existing value?

**Review by jinghaomiao — DISMISSED** [review]

File: modules/perception/onboard/component/trafficlights_perception_component.cc (hunk 3 above)
Comment (jinghaomiao) [reply modules/perception/onboard/component/trafficlights_perception_component.cc:91]: Please work with @changsh726 and @lfcarol  for an efficient use of it.

File: modules/perception/onboard/component/BUILD [hunk 4]
```diff
@@ -3,6 +3,55 @@ load("//tools:cpplint.bzl", "cpplint")
...
+    name = "perception_component_inner_camera",
+    srcs = [
+        "trafficlights_perception_component.cc",
+    ],
+    hdrs = [
+        "trafficlights_perception_component.h",
+    ],
+    copts = [
+        '-DMODULE_NAME=\\"perception\\"',
+    ],
```
Comment (storypku) [inline modules/perception/onboard/component/BUILD:23]: Define a `PERCEPTION_COPTS = [  '-DMODULE_NAME=\\"perception\\"' ]` line after the `package(...)` line, 

and use it in this BUILD file where needed:

```
copts = PERCEPTION_COPTS,
```

File: modules/perception/onboard/component/BUILD [hunk 5]
```diff
@@ -3,6 +3,55 @@ load("//tools:cpplint.bzl", "cpplint")
...
+    linkopts = ["-lstdc++"],
+    deps = [
```
Comment (storypku) [inline modules/perception/onboard/component/BUILD:25]: You don't need to link `stdc++`

File: modules/perception/onboard/component/BUILD [hunk 6]
```diff
@@ -3,6 +3,55 @@ load("//tools:cpplint.bzl", "cpplint")
...
+        "//modules/perception/onboard/transform_wrapper",
+        "//modules/perception/proto:traffic_light_detection_cc_proto",
+        "//modules/transform/proto:transform_cc_proto",
+        "//modules/v2x/common:v2x_proxy_gflags",
+        "//modules/v2x/proto:v2x_traffic_light_cc_proto",
+        "@eigen",
+        "@opencv//:core",
+        "@opencv//:highgui",
+        "@opencv//:imgcodecs",
+        "@opencv//:imgproc",
```
Comment (storypku) [inline modules/perception/onboard/component/BUILD:50]: Do you really to depend on all core libraries of OpenCV. 
Don't over estimate your dependencies.
Generally speaking, only those targets to which the included headers belong should be specified in the `deps` section.

File: modules/perception/onboard/component/trafficlights_perception_component.cc [hunk 7]
```diff
@@ -1058,8 +1030,8 @@ void TrafficLightsPerceptionComponent::SyncV2XTrafficLights(
       if (std::fabs(camera_frame_timestamp - v2x_timestamp) <
           v2x_sync_interval_seconds_) {
         const int v2x_lights_num =
-            (*itr).current_lane_trafficlight().single_traffic_light_size();
-        const auto& v2x_lights = (*itr).current_lane_trafficlight();
+            (*itr).road_traffic_light(0).single_traffic_light_size();
+        const auto& v2x_lights = (*itr).road_traffic_light(0);
         for (int i = 0; i < v2x_lights_num; ++i) {
```
Comment (storypku) [inline modules/perception/onboard/component/trafficlights_perception_component.cc:1035]: Use `itr->` rather than `(*iter).` to improve readability.

**Review by storypku — CHANGES_REQUESTED** [review]

File: modules/perception/onboard/component/trafficlights_perception_component.cc (hunk 3 above)
Comment (lfcarol) [reply modules/perception/onboard/component/trafficlights_perception_component.cc:91]: > make_shared may cause Eigen error @lfcarol

Fix it by "frame_.reset(new camera::CameraFrame);" , "make_shared" will case crash.

Comment (lfcarol) [reply modules/perception/onboard/component/trafficlights_perception_component.cc:91]: > Please work with @changsh726 and @lfcarol for an efficient use of it.

OK

**Review by lx18233184051 — DISMISSED** [review]

File: modules/perception/onboard/component/trafficlights_perception_component.cc [hunk 8]
```diff
@@ -39,6 +40,9 @@
 namespace apollo {
 namespace perception {
 namespace onboard {
+
+using ::apollo::cyber::Clock;
+
 using TLCamID = apollo::perception::TrafficLightDetection::CameraID;
 using apollo::cyber::common::GetAbsolutePath;
```
Comment (changsh726) [inline modules/perception/onboard/component/trafficlights_perception_component.cc:47]: Better to put `using apollo::cyber::Clock;` here.

Comment (xiaoMrzhang) [reply modules/perception/onboard/component/trafficlights_perception_component.cc:47]: OK

Comment (changsh726) [reply modules/perception/onboard/component/trafficlights_perception_component.cc:47]: Actually what I mean is put `using apollo::cyber::Clock` on the next line of `using::apollo::cyber::common::GetAbsolutePath`...

Comment (xiaoMrzhang) [reply modules/perception/onboard/component/trafficlights_perception_component.cc:47]: get it

**Review by changsh726 — APPROVED** [review]

**Pull request merged**